In [1]:
import pandas as pd

df = pd.read_csv("sales_transactions.csv")
print(df)


    account           name  order       sku  quantity  unit price  ext price
0    383080       Will LLC  10001  B1-20000         7       33.69     235.83
1    383080       Will LLC  10001  S1-27722        11       21.12     232.32
2    383080       Will LLC  10001  B1-86481         3       35.99     107.97
3    412290  Jerde-Hilpert  10005  S1-06532        48       55.82    2679.36
4    412290  Jerde-Hilpert  10005  S1-82801        21       13.62     286.02
5    412290  Jerde-Hilpert  10005  S1-06532         9       92.55     832.95
6    412290  Jerde-Hilpert  10005  S1-47412        44       78.91    3472.04
7    412290  Jerde-Hilpert  10005  S1-27722        36       25.42     915.12
8    218895      Kulas Inc  10006  S1-27722        32       95.66    3061.12
9    218895      Kulas Inc  10006  B1-33087        23       22.55     518.65
10   218895      Kulas Inc  10006  B1-33364         3       72.30     216.90
11   218895      Kulas Inc  10006  B1-20000        -1       72.18     -72.18

- You can see in the data that the file contains 3 different orders (order --> 10001, 10005 and 10006) and that each order consists has multiple products (aka skus).

- The question we would like to answer is: “What percentage of the order total does each sku represent?”

- For example, if we look at order 10001 with a total of $576.12, the break down would be:

- - B1-20000 = $235.83 or 40.9%

- - S1-27722 = $232.32 or 40.3%

- - B1-86481 = $107.97 or 18.7%

- The tricky part in this calculation is that we need to get a total for each order and combine it back with the transaction level detail in order to get the percentages.

In [4]:
# using group by

Order_Total = df.groupby('order')["ext price"].sum()
print(Order_Total)

order_total = df.groupby('order')["ext price"].sum().rename("Order_Total").reset_index()
df_1 = df.merge(order_total)
df_1["Percent_of_Order"] = df_1["ext price"] / df_1["Order_Total"]

print(df_1)

order
10001     576.12
10005    8185.49
10006    3724.49
Name: ext price, dtype: float64
    account           name  order       sku  quantity  unit price  ext price  \
0    383080       Will LLC  10001  B1-20000         7       33.69     235.83   
1    383080       Will LLC  10001  S1-27722        11       21.12     232.32   
2    383080       Will LLC  10001  B1-86481         3       35.99     107.97   
3    412290  Jerde-Hilpert  10005  S1-06532        48       55.82    2679.36   
4    412290  Jerde-Hilpert  10005  S1-82801        21       13.62     286.02   
5    412290  Jerde-Hilpert  10005  S1-06532         9       92.55     832.95   
6    412290  Jerde-Hilpert  10005  S1-47412        44       78.91    3472.04   
7    412290  Jerde-Hilpert  10005  S1-27722        36       25.42     915.12   
8    218895      Kulas Inc  10006  S1-27722        32       95.66    3061.12   
9    218895      Kulas Inc  10006  B1-33087        23       22.55     518.65   
10   218895      Kulas Inc  100

## **Second Approach - Using Transform**
Using the original data, let’s try using **transform** and **groupby** and see what we get:

In [8]:
df['Order_Total'] = df.groupby('order')["ext price"].transform('sum')
print(df['Order_Total'])

0      576.12
1      576.12
2      576.12
3     8185.49
4     8185.49
5     8185.49
6     8185.49
7     8185.49
8     3724.49
9     3724.49
10    3724.49
11    3724.49
Name: Order_Total, dtype: float64


In [10]:
df["Percent_of_Order"] = df["ext price"] / df["Order_Total"]
Percent_of_Order = df["Percent_of_Order"]
print(Percent_of_Order)
print(df)

0     0.409342
1     0.403249
2     0.187409
3     0.327330
4     0.034942
5     0.101759
6     0.424170
7     0.111798
8     0.821890
9     0.139254
10    0.058236
11   -0.019380
Name: Percent_of_Order, dtype: float64
    account           name  order       sku  quantity  unit price  ext price  \
0    383080       Will LLC  10001  B1-20000         7       33.69     235.83   
1    383080       Will LLC  10001  S1-27722        11       21.12     232.32   
2    383080       Will LLC  10001  B1-86481         3       35.99     107.97   
3    412290  Jerde-Hilpert  10005  S1-06532        48       55.82    2679.36   
4    412290  Jerde-Hilpert  10005  S1-82801        21       13.62     286.02   
5    412290  Jerde-Hilpert  10005  S1-06532         9       92.55     832.95   
6    412290  Jerde-Hilpert  10005  S1-47412        44       78.91    3472.04   
7    412290  Jerde-Hilpert  10005  S1-27722        36       25.42     915.12   
8    218895      Kulas Inc  10006  S1-27722        32       9